csa.rs

```diff
@@ -0,0 +1,134 @@
+use std::io;
+
+const MAX_STATIONS: usize = 100000;
+
+#[derive(Debug)]
+struct Connection {
+    departure_station: usize,
+    arrival_station: usize,
+    departure_timestamp: u32,
+    arrival_timestamp: u32
+}
+
+impl Connection {
+    fn parse(line: String) -> Connection {
+        let splitted = line.split(" ")
+            .map(|bite| {
+                bite.parse()
+                    .ok()
+                    .expect(&format!("failed to read {} as integer", bite))
+            })
+            .collect::<Vec<u32>>();
+        
+        Connection {
+            departure_station: splitted[0] as usize,
+            arrival_station: splitted[1] as usize,
+            departure_timestamp: splitted[2],
+            arrival_timestamp: splitted[3]
+        }
+    }
+}
+
+fn csa_main_loop(timetable: &Vec<Connection>, arrival_station: usize, earliest_arrival: &mut [u32], in_connection: &mut [usize]) {
+    let mut earliest = std::u32::MAX;
+        
+    for (i, connection) in timetable.iter().enumerate() {
+        if connection.departure_timestamp >= earliest_arrival[connection.departure_station] &&
+                connection.arrival_timestamp < earliest_arrival[connection.arrival_station] {
+            earliest_arrival[connection.arrival_station] = connection.arrival_timestamp;
+            in_connection[connection.arrival_station] = i as usize;
+            
+            if connection.arrival_station == arrival_station && connection.arrival_timestamp < earliest {
+                earliest = connection.arrival_timestamp;
+            }
+        } else if connection.arrival_timestamp > earliest {
+            return;                    
+        }                    
+    }
+}
+
+fn csa_print_result(timetable: &Vec<Connection>, in_connection: &[usize; MAX_STATIONS], arrival_station: usize) {
+    if in_connection[arrival_station] == std::u32::MAX as usize {
+        println!("NO_SOLUTION");
+    } else {
+        let mut route = Vec::new();
+        let mut last_connection_index = in_connection[arrival_station];
+        
+        while last_connection_index != std::u32::MAX as usize {
+            let ref connection = timetable[last_connection_index];
+            route.push(connection);
+            last_connection_index = in_connection[connection.departure_station];
+        }
+        
+        for connection in route.iter().rev() {
+            println!("{} {} {} {}", connection.departure_station, connection.arrival_station, connection.departure_timestamp, connection.arrival_timestamp); 
+        }
+    }           
+    println!("");
+}
+
+fn csa_compute(timetable: &Vec<Connection>, departure_station: usize, arrival_station: usize, departure_time: u32)
+{ 
+    let mut in_connection = [std::u32::MAX as usize; MAX_STATIONS];
+    let mut earliest_arrival = [std::u32::MAX; MAX_STATIONS];
```

Comment:
It will be allocated on the stack, it may be a bit too much if `MAX_STATIONS` is too big. You can just use

```rust
let mut v = Vec::new();
v.extend(repeat(std::u32::MAX).take(MAX_STATIONS));// damn, there was a Vec::from_element(elem, nb) before...
```